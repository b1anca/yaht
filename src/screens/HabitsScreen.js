import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import NavLink from "../components/NavLink";
import { fetchHabits } from "../features/habits/habitsActions";
import withSideNavLayout from "../hoc/withSidenavLayout";
import { H1 } from "../components/Typography";

const HabitsScreen = () => {
  const { habits } = useSelector((state) => state.habits);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchHabits());
  }, []);

  return (
    <>
      <H1>Your habits</H1>
      {habits.map((habit, index) => (
        <div
          key={index}
          className="text-sm text-slate-700 mb-4 rounded-lg bg-slate-100 px-6 py-5"
        >
          {habit.name}
        </div>
      ))}
      <NavLink primary to="/habits/new">
        Create habit
      </NavLink>
    </>
  );
};

HabitsScreen.displayName = "HabitsScreen";

export default withSideNavLayout(HabitsScreen);
