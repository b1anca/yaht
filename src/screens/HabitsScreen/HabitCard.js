import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";
import { H2, P } from "../../components/Typography";

const HabitCard = ({ id, name, overall_progress }) => (
  <NavLink to={`/habits/${id}`} className="basis-1/4">
    <div className="p-4 m-1 rounded hover:text-slate-100 bg-slate-300/5">
      <H2>{name}</H2>
      <div className="flex justify-between">
        <div className="flex-col flex-1">
          <P>Overall progress</P>
          {/* TODO: completion rate */}
          <P className="!text-4xl">{overall_progress}%</P>
        </div>
        <div className="flex-col flex-1">
          <P>Days in a row</P>
          <P bold>02</P>
          <P>Record</P>
          <P bold>02</P>
        </div>
      </div>
    </div>
  </NavLink>
);

HabitCard.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  overall_progress: PropTypes.string,
};

export default HabitCard;
