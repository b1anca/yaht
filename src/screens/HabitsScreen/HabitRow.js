import React, { useCallback, useState } from "react";
import { useDispatch } from "react-redux";
import classNames from "classnames";
import PropTypes from "prop-types";
import { isSameDay } from "date-fns";
import { deleteTask, createTask } from "../../features/tasks/taskActions";
import { DEFAULT_HABIT_COLOR } from "../../constants";
import NavLink from "../../components/NavLink";
import { P } from "../../components/Typography";

const Day = ({ task, day, habitId, habitColor }) => {
  const currentDate = new Date();
  const dispatch = useDispatch();
  const [loading, setLoading] = useState();

  const handleDayTaskClick = useCallback(() => {
    if (loading || day > currentDate) {
      return;
    }

    setLoading(true);
    if (task) {
      dispatch(deleteTask({ habitId, id: task.id })).then(() =>
        setLoading(false)
      );
    } else {
      dispatch(createTask({ habitId, completed_at: day.toISOString() })).then(
        () => setLoading(false)
      );
    }
  }, [day, currentDate, habitId, loading]);

  return (
    <div
      title={
        day > currentDate
          ? "This is a future date. You can't mark tasks as completed in advance."
          : task
          ? "Click to unmark this day."
          : "Click to mark this day."
      }
      style={{
        backgroundColor:
          task || loading ? habitColor || DEFAULT_HABIT_COLOR : "unset",
      }}
      className={classNames(
        "m-auto hover:border-2 whitespace-nowrap rounded w-8 h-8 items-center inline-flex justify-center",
        {
          "cursor-pointer hover:border-zinc-500": day <= currentDate,
          "animate-pulse": loading,
          "dark:!bg-[#24292E] !bg-slate-200": !(task || loading),
        }
      )}
      onClick={handleDayTaskClick}
    />
  );
};

const HabitRow = ({
  id,
  name,
  tasks,
  days,
  color,
  current_streak,
  record_streak,
  overall_progress,
}) => {
  return (
    <>
      <NavLink
        to={`/habits/${id}`}
        tetriary
        className="!p-0 truncate absolute sm:static mr-2 !text-base"
      >
        {name}
      </NavLink>
      {days.map((day) => {
        const completedTaskForTheDay = tasks.find((task) =>
          isSameDay(new Date(task.completed_at), day)
        );

        return (
          <Day
            key={`day-${day.toISOString()}`}
            day={day}
            task={completedTaskForTheDay}
            habitId={id}
            habitColor={color}
          />
        );
      })}
      <P className="!mb-0 text-center">{current_streak}</P>
      <P className="!mb-0 text-center">{record_streak}</P>
      <P className="!mb-0 text-center">{overall_progress}%</P>
    </>
  );
};

Day.propTypes = {
  day: PropTypes.object,
  task: PropTypes.object,
  habitId: PropTypes.number,
  habitColor: PropTypes.string,
};

HabitRow.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  color: PropTypes.string,
  tasks: PropTypes.array,
  days: PropTypes.array,
  current_streak: PropTypes.number,
  record_streak: PropTypes.number,
  overall_progress: PropTypes.string,
};

export default HabitRow;
