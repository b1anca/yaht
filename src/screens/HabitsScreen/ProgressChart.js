import React from "react";
import PropTypes from "prop-types";
import { PieChart, Pie, ResponsiveContainer, Cell, Tooltip } from "recharts";
import { COLORS, DEFAULT_HABIT_COLOR } from "../../constants";
import { P } from "../../components/Typography";

const ProgressChart = ({ value, color, withTooltip = false }) => {
  const data = [
    { name: "Overall progress", value },
    { name: "Uncompleted tasks", value: Number((100 - value).toFixed(2)) },
  ];

  return (
    <div className="h-[130px] w-[130px] relative">
      <P
        bold
        className="absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2"
      >
        {value}%
      </P>
      <ResponsiveContainer width="100%" height="100%">
        <PieChart>
          <Pie
            data={data}
            innerRadius={40}
            outerRadius={60}
            paddingAngle={5}
            dataKey="value"
            stroke={COLORS.slate400}
          >
            {data.map((_, index) => (
              <Cell
                key={`cell-${index}`}
                fill={
                  index === 0 ? color || DEFAULT_HABIT_COLOR : "transparent"
                }
              />
            ))}
          </Pie>
          {withTooltip && <Tooltip />}
        </PieChart>
      </ResponsiveContainer>
    </div>
  );
};

ProgressChart.propTypes = {
  value: PropTypes.number,
  color: PropTypes.string,
  withTooltip: PropTypes.bool,
};

export default ProgressChart;
