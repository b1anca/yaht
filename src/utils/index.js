export const getRangeFromViewportWidth = () => {
  const viewportWidth = window.innerWidth;

  if (viewportWidth <= 390) {
    return 3;
  } else if (viewportWidth <= 500) {
    return 4;
    // sm
  } else if (viewportWidth <= 640) {
    return 6;
    // md
  } else if (viewportWidth <= 768) {
    return 8;
    // lg
  } else if (viewportWidth <= 1024) {
    return 9;
  } else {
    return 10;
  }
};
