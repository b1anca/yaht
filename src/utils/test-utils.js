import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { setupStore } from "../app/store";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }
  return {
    store,
    user: userEvent.setup(),
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
