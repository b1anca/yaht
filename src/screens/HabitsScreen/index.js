import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import classNames from "classnames";
import PropTypes from "prop-types";
import { H2, P } from "../../components/Typography";
import { fetchHabits } from "../../features/habits/habitActions";
import {
  getDaysInMonth,
  areDatesEqual,
  monthNames,
} from "../../utils/dateHelpers";
import HabitRow from "./HabitRow";
import HabitCard from "./HabitCard";

const DayHeader = ({ day, currentDate }) => (
  <th>
    <P
      className={classNames({
        "text-slate-400 font-normal": !areDatesEqual(day, currentDate),
        "text-slate-200 text-base !mb-0": areDatesEqual(day, currentDate),
      })}
    >
      {day.getDate()}
    </P>
    {areDatesEqual(day, currentDate) && (
      <P className="text-slate-200">{monthNames[day.getMonth()]}</P>
    )}
  </th>
);

const HabitsScreen = () => {
  const currentDate = new Date();
  const { habits } = useSelector((state) => state.habits);
  const dispatch = useDispatch();
  const days = getDaysInMonth(
    currentDate.getMonth(),
    currentDate.getFullYear()
  );

  useEffect(() => {
    dispatch(fetchHabits());
  }, []);

  return (
    <div>
      <H2>Habits tracker</H2>
      <div className="bg-slate-400/5 rounded p-6">
        <table>
          <thead>
            <tr>
              <th />
              {days.map((day) => (
                <DayHeader
                  key={day.toISOString()}
                  day={day}
                  currentDate={currentDate}
                />
              ))}
            </tr>
          </thead>
          <tbody>
            {habits.map((habit) => (
              <HabitRow key={habit.id} {...habit} days={days} />
            ))}
          </tbody>
        </table>
      </div>
      <H2 className="mt-16">Your habits</H2>
      <div className="flex flex-wrap">
        {habits.map((habit) => (
          <HabitCard key={habit.id} {...habit} />
        ))}
      </div>
    </div>
  );
};

DayHeader.propTypes = {
  day: PropTypes.object,
  currentDate: PropTypes.object,
};

export default HabitsScreen;
