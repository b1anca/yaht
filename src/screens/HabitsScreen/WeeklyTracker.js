import React, { useEffect, useState, useMemo } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";
import { format, isSameDay } from "date-fns";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faArrowDown,
  faArrowUp,
  faChevronLeft,
  faChevronRight,
} from "@fortawesome/free-solid-svg-icons";
import { COLORS, TEXT_COLORS } from "../../constants";
import HabitRow from "./HabitRow";
import { getWeeklyDateRange, getDaysInMonth } from "../../utils/dateHelpers";
import { Heading, P } from "../../components/Typography";
import ProgressBar from "../../components/ProgressBar";

const DayHeader = ({ day }) => {
  const isCurrentDate = isSameDay(day, new Date());

  return (
    <div className="grow">
      <P
        className={classNames("!mb-0 p-1 text-center", {
          [TEXT_COLORS.secondary]: !isCurrentDate,
          "font-bold": isCurrentDate,
        })}
      >
        {day.getDate()} <br />
        {format(day, "EEE")}
      </P>
    </div>
  );
};

const WeeklyTracker = ({ habits, data }) => {
  const [dateRange, setDateRange] = useState([]);
  const currentDate = new Date();

  useEffect(() => {
    setDateRange(getDaysInMonth(currentDate));
  }, []);

  // const onClickLeft = () => {
  //   const date = dateRange[0];
  //   date.setDate(date.getDate() - 1);
  //   setDateRange(getWeeklyDateRange(date));
  // };

  // const onClickRight = () => {
  //   const date = dateRange[6];
  //   date.setDate(date.getDate() + 1);
  //   setDateRange(getWeeklyDateRange(date));
  // };

  // const lastWeekSum = useMemo(() => {
  //   const startDate = new Date(dateRange[0]);

  //   return Array.from({ length: 7 }).reduce((sum, _, i) => {
  //     const date = new Date(startDate);
  //     date.setDate(startDate.getDate() - i - 1);
  //     return sum + (data[date.toLocaleDateString()] || 0);
  //   }, 0);
  // }, [habits, dateRange]);

  // const currentWeekSum = useMemo(() => {
  //   const startDate = new Date(dateRange[0]);

  //   return Array.from({ length: 7 }).reduce((sum, _, i) => {
  //     const date = new Date(startDate);
  //     date.setDate(startDate.getDate() + i);
  //     return sum + (data[date.toLocaleDateString()] || 0);
  //   }, 0);
  // }, [habits, dateRange]);

  // const completedWeekPercentage = habits.length
  //   ? (currentWeekSum / (habits.length * 7)) * 100
  //   : 0;

  // const comparedToLastWeek = () => {
  //   const percentageChange = (
  //     lastWeekSum === 0
  //       ? 0
  //       : ((currentWeekSum - lastWeekSum) / lastWeekSum) * 100
  //   ).toFixed(2);

  //   let displayMessage, icon;
  //   if (percentageChange > 0) {
  //     displayMessage = `Up ${percentageChange}% from last week`;
  //     icon = faArrowUp;
  //   } else if (percentageChange < 0) {
  //     displayMessage = `Down ${Math.abs(percentageChange)}% from last week`;
  //     icon = faArrowDown;
  //   } else {
  //     displayMessage = "No change from last week";
  //   }

  //   return (
  //     <>
  //       {icon && (
  //         <FontAwesomeIcon
  //           icon={icon}
  //           style={{
  //             color: percentageChange > 0 ? COLORS.lime500 : COLORS.red500,
  //           }}
  //         />
  //       )}
  //       <P semibold className="!mb-0 ml-1">
  //         {displayMessage}
  //       </P>
  //     </>
  //   );
  // };

  return (
    <div className="border border-slate-900/10 dark:border-slate-100/10 p-3 rounded-sm mb-4">
      {/* <div className="flex items-center mt-6 mb-2">
        <div
          onClick={onClickLeft}
          className="border dark:border-zinc-200 rounded-full px-3 py-1 hover:cursor-pointer hover:bg-slate-100 dark:hover:bg-zinc-700 mr-4"
        >
          <FontAwesomeIcon
            size="sm"
            icon={faChevronLeft}
            className="dark:text-zinc-300 text-zinc-600"
          />
        </div>
        <div
          onClick={onClickRight}
          className="border dark:border-zinc-200 rounded-full px-3 py-1 hover:cursor-pointer hover:bg-slate-100 dark:hover:bg-zinc-700 mr-4"
        >
          <FontAwesomeIcon
            size="sm"
            icon={faChevronRight}
            className="dark:text-zinc-300 text-zinc-600"
          />
        </div>
        <Heading level="h3" className="!mb-0">
          {dateRange[0] && format(dateRange[0], "EEE MM/dd")} -{" "}
          {dateRange[6] && format(dateRange[6], "EEE MM/dd")}
        </Heading>
      </div>
      <ProgressBar value={completedWeekPercentage} className="mb-1" /> */}
      {/* <div className="flex justify-between mb-4 items-center">
        <div className="flex items-center">{comparedToLastWeek()}</div>
        <P semibold className="!mb-0">
          {completedWeekPercentage.toFixed(2)}% achieved
        </P>
      </div> */}
      {habits.length > 0 ? (
        <div>
          <div className="flex mb-2">
            <div className="w-[200px]" />
            {dateRange.map((day) => (
              <DayHeader key={day.toISOString()} day={day} />
            ))}
            {/* <P className="!mb-0 p-1 text-center font-semibold">
              current <br />
              streak
            </P>
            <P className="!mb-0 p-1 text-center font-semibold">
              longest <br />
              streak
            </P>
            <P className="!mb-0 p-1 text-center font-semibold">
              overall <br />
              progress
            </P> */}
          </div>
          {habits.map((habit) => (
            <HabitRow key={habit.id} {...habit} days={dateRange} />
          ))}
        </div>
      ) : (
        <P>
          Looks like you&apos;re just getting started! Click on the &apos;Add
          Habit&apos; button to begin your journey towards positive change.
          Let&apos;s create your first habit today!
        </P>
      )}
    </div>
  );
};

DayHeader.propTypes = {
  day: PropTypes.object,
};

WeeklyTracker.propTypes = {
  habits: PropTypes.array,
  data: PropTypes.object,
};

export default WeeklyTracker;
