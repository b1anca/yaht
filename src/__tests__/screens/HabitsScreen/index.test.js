import React from "react";
import { screen, waitFor } from "@testing-library/react";
import { renderWithProviders } from "../../../utils/test-utils";
import HabitsScreen from "../../../screens/HabitsScreen";

jest.useFakeTimers().setSystemTime(new Date("2023-01-10"));

global.ResizeObserver = require("resize-observer-polyfill");

describe("HabitsScreen", () => {
  it("should render correctly", async () => {
    const { asFragment } = renderWithProviders(<HabitsScreen />);

    await waitFor(() => {
      screen.getAllByText(/reading/i);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
