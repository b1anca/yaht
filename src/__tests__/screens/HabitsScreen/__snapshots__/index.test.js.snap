// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitsScreen should render correctly 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="mb-4 text-2xl font-extrabold tracking-tight"
    >
      Habits tracker
    </h2>
    <div
      class="bg-slate-400/5 rounded p-6"
    >
      <table>
        <thead>
          <tr>
            <th />
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-200 text-base !mb-0"
              >
                1
              </p>
              <p
                class="text-sm mb-2 tracking-tight text-slate-200"
              >
                Jun
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                2
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                3
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                4
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                5
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                6
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                7
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                8
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                9
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                10
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                11
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                12
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                13
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                14
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                15
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                16
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                17
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                18
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                19
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                20
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                21
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                22
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                23
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                24
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                25
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                26
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                27
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                28
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                29
              </p>
            </th>
            <th>
              <p
                class="text-sm mb-2 tracking-tight text-slate-400 font-normal"
              >
                30
              </p>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <p
                class="text-sm mb-2 tracking-tight whitespace-nowrap mt-2 mr-2"
              >
                <a
                  class="text-sm whitespace-nowrap py-2 px-6 rounded-full inline-flex justify-center rounded-lg font-semibold py-2.5 px-4 hover:text-slate-200 !p-0"
                  href="/habits/1"
                >
                  Reading
                </a>
              </p>
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center cursor-pointer"
                title="Click to mark this day."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                style="background-color: rgb(117, 171, 116);"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
            <td>
              <div
                class="border border-slate-500 hover:border-slate-200 whitespace-nowrap rounded w-6 h-6 items-center inline-flex justify-center"
                title="This is a future date. You can't mark tasks as completed in advance."
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <h2
      class="mb-4 text-2xl font-extrabold tracking-tight mt-16"
    >
      Your habits
    </h2>
    <div
      class="flex flex-wrap"
    >
      <a
        class="basis-1/4"
        href="/habits/1"
      >
        <div
          class="p-4 m-1 rounded hover:text-slate-100 bg-slate-300/5"
        >
          <h3
            class="mb-4 font-extrabold tracking-tight text-xl"
            level="h4"
          >
            Reading
          </h3>
          <div
            class="flex justify-between"
          >
            <div
              class="flex-col flex-1"
            >
              <div
                class="h-[130px] w-[130px] mr-4"
              >
                <div
                  class="recharts-responsive-container"
                  style="width: 100%; height: 100%; min-width: 0;"
                />
              </div>
            </div>
            <div
              class="flex-col flex-1"
            >
              <p
                class="text-sm mb-2 tracking-tight"
              >
                Days in a row
              </p>
              <p
                class="text-sm mb-2 tracking-tight font-extrabold"
              >
                02
              </p>
              <p
                class="text-sm mb-2 tracking-tight"
              >
                Record
              </p>
              <p
                class="text-sm mb-2 tracking-tight font-extrabold"
              >
                02
              </p>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
